### ISSUE TITLE
[DataGridPro] useGridSelector doesn't work

### ISSUE THREAD

**Issue by peppermint-juli** [mui-org/material-ui-x#1560]
I'm trying to use the apiRef to get track of the columns that are hidden or shown. Right now the oly way is by using either `onStateChange`, which makes the app crash or the `apiRef`, which also makes the app crash but because of this error. Basically using the useGridSelector fails because apiRef is not created correctly with `useGridApiRef` hook.

- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯
![image](https://user-images.githubusercontent.com/27710492/116939297-bc739b00-ac31-11eb-87f5-b3b1a0d26e75.png)

## Expected Behavior 🤔
No crashing 

## Steps to Reproduce 🕹
Here's a codesanbox that causes the error
https://codesandbox.io/s/material-demo-forked-6vvrb?file=/demo.tsx

## Context 🔦

Get track
## Your Environment 🌎
I'm using NextJs in my webapp, but the CodeSandBox uses a pure React app so it fails with both. Both using Chrome.

## Order id 💳
 # 18479

**Comment by dtassone** [comment]
FYI, you can't use the selector before the `apiRef` is fully created which happens after the component renders for the first time.
Also you have to pass `ApiRef` in `XGrid` as a prop

**Comment by oliviertassinari** [comment]
Is `useGridApiRef` even public? I thought it was a private API.

**Comment by oliviertassinari** [comment]
@peppermint-juli is this a duplicate of #820?

**Comment by dtassone** [comment]
> Is `useGridApiRef` even public? I thought it was a private API, we don't document it.

How do users use `apiRef` without `useGridApiRef`? 
https://material-ui.com/components/data-grid/rows/#apiref

**Comment by oliviertassinari** [comment]
@dtassone Ok, so `useGridApiRef` is documented and public. What about `useGridSelector`?

> How do users use apiRef without useGridApiRef?

With a React useRef, I don't understand the value of `useGridApiRef`: https://codesandbox.io/s/material-demo-forked-4ipje

```jsx
export default function ApiRefRowsGrid() {
  const apiRef = React.useRef<GridApi>();

  React.useEffect(() => {
    const subscription = interval(200).subscribe(() => {
      apiRef.current.updateRows([
        {
          id: randomInt(1, 4),
          username: randomUserName(),
          age: randomInt(10, 80)
        },
        {
          id: randomInt(1, 4),
          username: randomUserName(),
          age: randomInt(10, 80)
        }
      ]);
    });

    return () => {
      subscription.unsubscribe();
    };
  }, [apiRef]);

  return (
    <div style={{ height: 400, width: "100%" }}>
      <XGrid rows={rows} columns={columns} apiRef={apiRef} />
    </div>
  );
}

```

**Comment by m4theushw** [comment]
Fixed by #3431 

There's a warning now if `useGridSelector` is used before the state has been initialized.
